### PR TITLE
Add uint8, uint16, and int16 support for indexing operations

### PR DESCRIPTION
**Pull request by AmitMY**
## Summary

This PR extends index type support for PyTorch indexing operations to include **uint8, uint16, and int16** in addition to the existing int32 and int64, enabling memory-efficient indexing.

## Motivation

Supporting smaller integer types for indices allows:
- **Memory savings**: uint8 uses 8x less memory than int64 for small vocabularies
- **Better cache utilization**: More indices fit in cache, improving performance
- **Flexibility**: uint16 supports vocabularies up to 65K, int16 for signed use cases

### Core Implementation
- Updated `AT_DISPATCH_INDEX_TYPES` to include Short (int16) and UInt16
- Updated checks in `index_select`, `gather`, and `embedding` operations
- Added 32 EmbeddingBag template instantiations for new types
- Support across CPU, CUDA, and MPS backends

### Affected Operations
✅ `torch.embedding()` - CPU, CUDA, MPS
✅ `torch.index_select()` - CPU, CUDA, MPS  
✅ `torch.gather()` - CPU, CUDA
❌ `torch.take()` - Still int64 only (unchanged)

### Tests
- Added parametrized test `test_embedding_small_index_dtypes` (3 variants)
- Enhanced `test_index_select` to test 5 index types
- Added parametrized test `test_gather_small_index_dtypes` (12 variants)
- All tests passing ✅

## Example Usage

```python
import torch
import torch.nn as nn

# Embedding with uint8 indices (memory-efficient for small vocabularies)
embedding = nn.Embedding(200, 128)
indices_uint8 = torch.tensor([0, 5, 10], dtype=torch.uint8)
output = embedding(indices_uint8)  # Works!

# Index select with uint16 indices
x = torch.randn(1000, 512)
indices_uint16 = torch.tensor([0, 100, 500], dtype=torch.uint16)
result = torch.index_select(x, 0, indices_uint16)  # Works!

# Gather with int16 indices
x = torch.randn(10, 20)
indices_int16 = torch.tensor([[0, 5], [3, 8]], dtype=torch.int16)
result = torch.gather(x, 1, indices_int16)  # Works!
```

## Testing

All tests pass:
```bash
pytest test/nn/test_embedding.py -k test_embedding_small_index_dtypes -v  # 3 tests
pytest test/test_indexing.py -k test_index_select -v                      # 26 tests
pytest test/test_scatter_gather_ops.py -k test_gather_small_index_dtypes -v # 12 tests
```

Linting passes:
```bash
lintrunner -m upstream/main  # ✅ No lint issues
```

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Is this also for index_select?

And also for tensor indexing?

If so - some deprecation should be done as currently uint8 argument is casted as Boolean mask which is a very different, surprising, legacy semantics (which should be switched off IMO). So I'd suggest deprecating / switching off it first in the tensor indexing operator

**Comment by vadimkantorov** [comment]
and also scatter_reduce support for smaller index dtypes is needed for GNNs: https://github.com/pytorch/pytorch/issues/71300

**Comment by AmitMY** [comment]
Hi! I would love to go ahead with this

**Comment by Evidlo** [comment]
Can this be revived?  I have been wanting uint8 indexing support for years.  This will be a godsend for decreasing memory usage of my simulations.
